cmd/snap-update-ns/bootstrap.c

```diff
@@ -0,0 +1,129 @@
+int bootstrap_errno = 0;
+// bootstrap_msg contains a static string if something fails.
+const char* bootstrap_msg = NULL;
+
+// read_cmdline reads /proc/self/cmdline into the specified buffer, returning
+// number of bytes read.
+static ssize_t
+read_cmdline(char* buf, size_t buf_size)
+{
+    int fd = open("/proc/self/cmdline", O_RDONLY | O_CLOEXEC | O_NOFOLLOW);
+    if (fd < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot open /proc/self/cmdline";
+        return -1;
+    }
+    memset(buf, 0, buf_size);
+    ssize_t num_read;
+    if ((num_read = read(fd, buf, buf_size)) < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot read /proc/self/cmdline";
+    }
+    close(fd);
+    return num_read;
+}
+
+// find_snap_name scans the command line buffer and looks for the 1st argument.
+static const char*
+find_snap_name(char* buf, size_t buf_size, size_t num_read)
+{
+    // cmdline is an array of NUL ('\0') separated strings. We can skip over
+    // the first entry (program name) and look at the second entry, in our case
+    // it should be the snap name.
+    size_t argv0_len = strnlen(buf, buf_size);
+    if (argv0_len == num_read) {
+        return NULL;
+    }
+    return &buf[argv0_len + 1];
+}
+
+// setns_into_snap switches mount namespace into that of a given snap.
+static int
+setns_into_snap(const char* snap_name)
+{
+    // Construct the name of the .mnt file to open.
+    char buf[PATH_MAX];
+    int n = snprintf(buf, sizeof buf, "/run/snapd/ns/%s.mnt", snap_name);
+    if (n >= sizeof buf || n < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot format mount namespace file name";
+        return -1;
+    }
+
+    // Open the mount namespace file, note that we don't specify O_NOFOLLOW as
+    // that file is always a special, broken symbolic link.
+    int fd = open(buf, O_RDONLY | O_CLOEXEC);
+    if (fd < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot open mount namespace file";
+        return -1;
+    }
+
+    // Switch to the mount namespace of the given snap.
+    int err = setns(fd, CLONE_NEWNS);
+    if (err < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot switch mount namespace";
+    };
+
+    close(fd);
+    return err;
+}
+
+// bootstrap prepares snap-update-ns to work in the namespace of the snap given
+// on command line.
+void bootstrap(void)
+{
+    // We don't have argc/argv so let's imitate that by reading cmdline
+    char cmdline[1024];
+    memset(cmdline, 0, sizeof cmdline);
+    ssize_t num_read;
+    if ((num_read = read_cmdline(cmdline, sizeof cmdline)) < 0) {
+        return;
+    }
+
+    // Find the name of the snap by scanning the cmdline.  If there's no snap
+    // name given, just bail out. The go parts will scan this too.
+    const char* snap_name = find_snap_name(cmdline, sizeof cmdline, num_read);
+    if (*snap_name == '\0') {
```

Comment:
find_snap_name returns NULL.

Slightly worried about the rate of important errors in this logic. Please make sure to get a good second review.